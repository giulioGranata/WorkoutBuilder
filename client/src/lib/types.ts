export type WorkoutType = "recovery" | "endurance" | "tempo" | "threshold" | "vo2max" | "anaerobic";
export type Difficulty = "easy" | "standard" | "hard";

export type Step = {
  minutes: number;
  intensityPct: number;
  description: string;
  phase?: "warmup" | "work" | "recovery" | "cooldown";
};

export type Workout = {
  title: string;
  steps: Step[];
  totalMinutes: number;
  workMinutes?: number;
  recoveryMinutes?: number;
  avgIntensity?: number;
};

export interface WorkoutFormData {
  ftp: number;
  durationMin: number;
  type: WorkoutType;
  difficulty: Difficulty;
}
