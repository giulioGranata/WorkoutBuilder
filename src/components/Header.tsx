"use client";

import type { Session } from "@supabase/supabase-js";
import { Dumbbell, Menu, X } from "lucide-react";
import type { Route } from "next";
import Link from "next/link";
import { usePathname } from "next/navigation";
import { useEffect, useMemo, useState } from "react";

import { useSupabase } from "@/components/SupabaseProvider";
import { ThemeToggle } from "@/components/ThemeToggle";
import { Button } from "@/components/ui/button";
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuTrigger,
  dropdownMenuItemClass,
} from "@/components/ui/dropdown-menu";
import { cn } from "@/lib/utils";

const NAV_LINKS: Array<{ label: string; href: Route }> = [
  { label: "Generator", href: "/" },
  { label: "Pro", href: "/pro" },
];

function getDisplayName(session: Session | null) {
  const metadata = (session?.user?.user_metadata ?? {}) as {
    full_name?: string;
    user_name?: string;
  };

  return metadata.full_name ?? metadata.user_name ?? session?.user?.email ?? "";
}

function getInitials(input: string) {
  if (!input) return "WG";

  const tokens = input
    .split(/[\s@._-]+/g)
    .map((token) => token.trim())
    .filter(Boolean);

  if (tokens.length === 0) {
    return input.slice(0, 2).toUpperCase();
  }

  if (tokens.length === 1) {
    return tokens[0].slice(0, 2).toUpperCase();
  }

  return `${tokens[0][0] ?? ""}${tokens[1][0] ?? ""}`.toUpperCase();
}

export function Header() {
  const { session } = useSupabase();
  const pathname = usePathname();
  const [menuOpen, setMenuOpen] = useState(false);
  const isProfile = pathname?.startsWith("/profile");

  useEffect(() => {
    setMenuOpen(false);
  }, [pathname]);

  const displayName = useMemo(() => getDisplayName(session), [session]);
  const initials = useMemo(() => getInitials(displayName), [displayName]);

  return (
    <header className="sticky top-0 z-40 border-b border-[--border] surface-glass">
      <div className="mx-auto flex h-16 w-full max-w-6xl items-center justify-between px-4 sm:px-6">
        <div className="flex items-center gap-8">
          <Link
            href="/"
            className="flex items-center gap-3 motion-safe:transition-opacity motion-safe:duration-200 hover:opacity-90"
          >
            <span className="grid h-10 w-10 place-items-center rounded-xl bg-[--accent] text-[--accent-foreground] shadow-[0_1px_0_0_rgba(255,255,255,0.08)]">
              <Dumbbell className="h-4 w-4" />
            </span>
            <span className="font-semibold tracking-tight text-[--text-primary]">
              Workout Generator
            </span>
          </Link>

          <nav className="hidden items-center gap-6 text-sm font-medium md:flex">
            {NAV_LINKS.map((item) => {
              const active = pathname === item.href;

              if (active && item.label !== "Pro") {
                return null;
              }

              return (
                <Link
                  key={item.href}
                  href={item.href}
                  aria-current={active ? "page" : undefined}
                  className={cn(
                    "text-[--text-secondary] motion-safe:transition-colors motion-safe:duration-200 hover:text-[--text-primary] focus-visible:outline-none focus-visible:text-[--text-primary]",
                    active
                      ? "text-sky-400 underline underline-offset-4 decoration-sky-400/40"
                      : undefined,
                  )}
                >
                  {item.label}
                </Link>
              );
            })}
          </nav>
        </div>

        <div className="flex items-center gap-3">
          <ThemeToggle />

          {session ? (
            <DropdownMenu className="hidden md:inline-flex">
              <DropdownMenuTrigger className="flex h-9 w-9 items-center justify-center rounded-full border border-[--border] bg-[color:var(--color-surface-muted)] text-sm font-medium uppercase text-[--text-primary] motion-safe:transition-colors motion-safe:duration-200 hover:bg-[color:var(--color-surface)] focus-ring btn-press">
                {initials}
              </DropdownMenuTrigger>
              <DropdownMenuContent className="border border-[--border] bg-[color:var(--card)] text-[--text-primary] backdrop-blur-sm shadow-xl">
                <div className="px-3 pb-2 pt-3">
                  <p className="text-xs uppercase tracking-wide text-[--text-tertiary]">
                    Signed in
                  </p>
                  <p className="truncate text-sm text-[--text-secondary]">
                    {displayName || session.user?.email}
                  </p>
                </div>
                <div className="my-1 h-px bg-[--border]" />
                {!isProfile ? (
                  <Link href="/profile" className={dropdownMenuItemClass}>
                    Profile
                  </Link>
                ) : null}
                <form action="/sign-out" method="post" className="contents">
                  <DropdownMenuItem type="submit">Sign out</DropdownMenuItem>
                </form>
              </DropdownMenuContent>
            </DropdownMenu>
          ) : (
            <Button
              asChild
              variant="primary"
              size="sm"
              className="hidden rounded-full md:inline-flex"
            >
              <Link href="/sign-in">Sign in</Link>
            </Button>
          )}

          <Button
            type="button"
            variant="ghost"
            size="icon"
            className="h-9 w-9 text-[--text-secondary] md:hidden"
            onClick={() => setMenuOpen((prev) => !prev)}
            aria-label="Toggle navigation"
            aria-expanded={menuOpen}
          >
            {menuOpen ? (
              <X className="h-5 w-5" />
            ) : (
              <Menu className="h-5 w-5" />
            )}
          </Button>
        </div>
      </div>

      <div className={cn("md:hidden", menuOpen ? "block" : "hidden")}>
        <div className="border-t border-[--border] surface-glass px-4 py-3">
          <nav className="flex flex-col gap-2">
            {NAV_LINKS.map((item) => {
              const active = pathname === item.href;

              if (active && item.label !== "Pro") {
                return null;
              }

              return (
                <Link
                  key={item.href}
                  href={item.href}
                  aria-current={active ? "page" : undefined}
                  className={cn(
                    "rounded-lg px-3 py-2 text-sm font-medium text-[--text-secondary] motion-safe:transition-colors motion-safe:duration-200 hover:bg-[color:var(--color-surface-muted)] hover:text-[--text-primary] focus-ring",
                    active
                      ? "text-sky-400 underline underline-offset-4 decoration-sky-400/40"
                      : undefined,
                  )}
                >
                  {item.label}
                </Link>
              );
            })}
          </nav>

          <div className="mt-3 border-t border-[--border] pt-3">
            {session ? (
              <div className="flex flex-col gap-2">
                {!isProfile ? (
                  <Link
                    href="/profile"
                    className="rounded-lg px-3 py-2 text-sm font-medium text-[--text-secondary] motion-safe:transition-colors motion-safe:duration-200 hover:bg-[color:var(--color-surface-muted)] hover:text-[--text-primary] focus-ring"
                  >
                    Profile
                  </Link>
                ) : null}
                <form action="/sign-out" method="post">
                  <Button
                    type="submit"
                    variant="ghost"
                    className="w-full justify-start text-[--text-secondary]"
                  >
                    Sign out
                  </Button>
                </form>
              </div>
            ) : (
              <Button
                asChild
                variant="primary"
                size="sm"
                className="w-full rounded-full"
              >
                <Link href="/sign-in">Sign in</Link>
              </Button>
            )}
          </div>
        </div>
      </div>
    </header>
  );
}

export default Header;
