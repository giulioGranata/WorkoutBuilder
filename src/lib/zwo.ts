import type { Workout } from "./types";
import { isRampStep } from "./types";

const xmlEscape = (s: string) =>
  s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;");

const clamp = (n: number, min: number, max: number) =>
  Math.max(min, Math.min(max, n));

// Convert a Workout to Zwift .zwo XML using SteadyState blocks.
// Power is a fraction of FTP with 2 decimals, clamped to [0.30, 1.60].
// Duration is floor(minutes * 60) seconds.
export function toZwoXml(
  workout: Workout & { ftp: number; biasPct?: number }
): string {
  const bias = workout.biasPct ?? 100;
  const title = xmlEscape(workout.title);
  const description = xmlEscape(`FTP ${workout.ftp}W â€¢ Bias ${bias}%`);

  const parts: string[] = [];
  parts.push(`<?xml version="1.0" encoding="UTF-8"?>`);
  parts.push(`<workout_file>`);
  parts.push(`  <author>Workout Generator 2.0</author>`);
  parts.push(`  <name>${title}</name>`);
  parts.push(`  <description>${description}</description>`);
  parts.push(`  <workout>`);

  for (const step of workout.steps) {
    const duration = Math.floor(step.minutes * 60);
    if (isRampStep(step)) {
      const fromRatio = workout.ftp > 0 ? step.from / workout.ftp : 0;
      const toRatio = workout.ftp > 0 ? step.to / workout.ftp : 0;
      const low = clamp(fromRatio, 0.3, 1.6).toFixed(2);
      const high = clamp(toRatio, 0.3, 1.6).toFixed(2);
      parts.push(
        `    <Ramp Duration="${duration}" PowerLow="${low}" PowerHigh="${high}" />`
      );
    } else {
      const ratioRaw = workout.ftp > 0 ? step.intensity / workout.ftp : 0;
      const ratioClamped = clamp(ratioRaw, 0.3, 1.6);
      const power = ratioClamped.toFixed(2);
      parts.push(`    <SteadyState Duration="${duration}" Power="${power}" />`);
    }
  }

  parts.push(`  </workout>`);
  parts.push(`</workout_file>`);
  return parts.join("\n");
}

