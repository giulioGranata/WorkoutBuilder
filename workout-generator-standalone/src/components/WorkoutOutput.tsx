import { useState } from "react";
import { Workout } from "../lib/types";

interface WorkoutOutputProps {
  workout: Workout | null;
}

export function WorkoutOutput({ workout }: WorkoutOutputProps) {
  const [isCopying, setIsCopying] = useState(false);

  const handleExportJSON = () => {
    if (!workout) return;
    
    const dataStr = JSON.stringify(workout, null, 2);
    const dataBlob = new Blob([dataStr], { type: 'application/json' });
    const url = URL.createObjectURL(dataBlob);
    
    const link = document.createElement('a');
    link.href = url;
    link.download = `${workout.title.replace(/[^a-zA-Z0-9]/g, '_')}.json`;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    URL.revokeObjectURL(url);
  };

  const handleCopyToClipboard = async () => {
    if (!workout) return;
    
    setIsCopying(true);
    
    const workoutText = `${workout.title}\n\n` +
      workout.steps.map((step, index) => 
        `${index + 1}. ${step.description}`
      ).join('\n') +
      `\n\nTotal: ${workout.totalMinutes}'`;
    
    try {
      await navigator.clipboard.writeText(workoutText);
      console.log("Workout copied to clipboard");
    } catch (error) {
      console.error("Failed to copy to clipboard:", error);
    } finally {
      setIsCopying(false);
    }
  };

  return (
    <div className="bg-gray-850 rounded-xl border border-gray-700 p-6">
      <div className="flex items-center justify-between mb-6">
        <div className="flex items-center">
          <div className="w-5 h-5 bg-emerald-500 rounded mr-3">
            <span className="block w-full h-full text-white text-xs font-bold flex items-center justify-center">#</span>
          </div>
          <h2 className="text-xl font-semibold text-white">Generated Workout</h2>
        </div>
        {workout && (
          <div className="flex items-center space-x-2">
            <button
              onClick={handleExportJSON}
              className="px-3 py-2 bg-gray-700 hover:bg-gray-600 text-gray-300 border border-gray-600 rounded-lg text-sm"
              title="Export as JSON"
            >
              â¬‡ JSON
            </button>
            <button
              onClick={handleCopyToClipboard}
              disabled={isCopying}
              className="px-3 py-2 bg-gray-700 hover:bg-gray-600 text-gray-300 border border-gray-600 rounded-lg text-sm"
              title="Copy to clipboard"
            >
              ðŸ“‹ Copy
            </button>
          </div>
        )}
      </div>

      {workout ? (
        <div className="workout-content">
          {/* Workout Title */}
          <div className="mb-6">
            <h3 className="text-2xl font-bold text-white mb-2">
              {workout.title}
            </h3>
            <div className="flex items-center text-sm text-gray-400">
              <span>ðŸ•’ Total: {workout.totalMinutes} minutes</span>
            </div>
          </div>

          {/* Workout Steps */}
          <div className="space-y-3">
            {workout.steps.map((step, index) => (
              <div
                key={index}
                className="bg-gray-700/50 rounded-lg p-4 border-l-4 border-emerald-500"
              >
                <div className="flex items-center justify-between mb-2">
                  <span className="text-emerald-400 text-xs font-medium uppercase tracking-wider">
                    Step {index + 1}
                  </span>
                  <span className="text-xs text-gray-500">
                    {step.minutes} minute{step.minutes !== 1 ? 's' : ''}
                  </span>
                </div>
                <div className="text-white font-medium mb-1">{step.intensityPct}% FTP</div>
                <div className="text-gray-300 text-sm">{step.description}</div>
              </div>
            ))}
          </div>

          {/* Export Actions */}
          <div className="mt-6 pt-6 border-t border-gray-700">
            <div className="flex flex-col sm:flex-row gap-3">
              <button
                onClick={handleExportJSON}
                className="flex-1 flex items-center justify-center px-4 py-3 bg-blue-600 hover:bg-blue-700 text-white font-medium rounded-lg transition-colors duration-200"
              >
                â¬‡ Export JSON
              </button>
              <button
                onClick={handleCopyToClipboard}
                disabled={isCopying}
                className="flex-1 flex items-center justify-center px-4 py-3 bg-gray-700 hover:bg-gray-600 text-white font-medium rounded-lg transition-colors duration-200 border border-gray-600"
              >
                ðŸ“‹ {isCopying ? "Copying..." : "Copy Text"}
              </button>
            </div>
          </div>
        </div>
      ) : (
        <div className="empty-state text-center py-12">
          <div className="text-4xl text-gray-600 mb-4">ðŸš´</div>
          <h3 className="text-lg font-medium text-gray-400 mb-2">No workout generated yet</h3>
          <p className="text-gray-500">Configure your settings and click "Generate Workout" to create a personalized training session.</p>
        </div>
      )}
    </div>
  );
}