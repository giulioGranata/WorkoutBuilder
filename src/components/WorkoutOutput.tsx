"use client";

import { Button } from "@/components/ui/button";
import { useToast } from "@/hooks/use-toast";
import { Step, Workout } from "@/lib/types";
import { getParamInt, setParam } from "@/lib/url";
import { toZwoXml } from "@/lib/zwo";
import {
  Bike,
  ChevronDown,
  Code,
  Download,
  FileCog,
  FileText,
  Info,
  ListOrdered,
  Minus,
  Plus,
} from "lucide-react";
import { useEffect, useMemo, useRef, useState } from "react";
import WorkoutChart from "./WorkoutChart";
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuTrigger,
} from "./ui/dropdown-menu";
import {
  Tooltip,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger,
} from "./ui/tooltip";

interface WorkoutOutputProps {
  workout: Workout | null;
  attempted?: boolean;
}

const BIAS_MIN = 75;
const BIAS_MAX = 125;

// --- helpers (bias) ---
export const clamp = (v: number, min: number, max: number) =>
  Math.max(min, Math.min(max, v));

export const applyBias = (watts: number, biasPct: number) =>
  Math.max(0, Math.round(watts * (biasPct / 100)));

export function WorkoutOutput({ workout, attempted = false }: WorkoutOutputProps) {
  const { toast } = useToast();
  const [menuOpen, setMenuOpen] = useState(false);

  const biasFromUrlRef = useRef(false);
  const [bias, setBias] = useState<number>(() => {
    if (typeof window !== "undefined") {
      const url = new URL(window.location.href);
      const b = getParamInt(url, "bias");
      if (b !== null && b >= BIAS_MIN && b <= BIAS_MAX) {
        biasFromUrlRef.current = true;
        return b;
      }
    }
    return 100;
  });
  const nudge = (delta: number) => {
    if (bias + delta > BIAS_MAX || bias + delta < BIAS_MIN) return;
    setBias((b) => clamp(b + delta, BIAS_MIN, BIAS_MAX));
  };

  useEffect(() => {
    if (typeof window === "undefined") return;
    if (!biasFromUrlRef.current && bias === 100) return;
    const url = new URL(window.location.href);
    setParam(url, "bias", clamp(bias, BIAS_MIN, BIAS_MAX));
    biasFromUrlRef.current = true;
  }, [bias]);

  // Compute a biased view of the workout (steps only)
  const biasedSteps = useMemo<Step[]>(
    () =>
      workout
        ? workout.steps.map((s) => ({
            ...s,
            intensity: applyBias(s.intensity, bias),
          }))
        : [],
    [workout, bias]
  );

  // Recompute avg intensity (W) using biased steps (duration-weighted)
  const biasedAvgIntensity = useMemo<number>(() => {
    if (!workout || biasedSteps.length === 0 || !workout.totalMinutes) return 0;
    const weighted =
      biasedSteps.reduce((sum, s) => sum + s.intensity * s.minutes, 0) /
      workout.totalMinutes;
    return Math.round(weighted);
  }, [workout, biasedSteps]);

  const normalizeDescription = (text: string) =>
    text.replace(/truncated/gi, "shortened");

  const buildWorkoutText = () => {
    if (!workout) return "";
    const header = `${workout.title} • FTP: ${workout.ftp} W • Bias: ${bias}%\n\n`;
    const body = biasedSteps
      .map(
        (step, index) =>
          `${index + 1}. ${step.minutes}' — ${
            step.intensity
          } W — ${normalizeDescription(step.description)}`
      )
      .join("\n");
    const footer = `\n\nTotal: ${workout.totalMinutes}'\nAvg: ${biasedAvgIntensity} W`;
    return header + body + footer;
  };

  const handleExportJSON = () => {
    if (!workout) return;

    const sanitizedSteps = biasedSteps.map((s) => ({
      ...s,
      description: normalizeDescription(s.description),
    }));

    const payload = {
      ...workout,
      steps: sanitizedSteps, // export with current bias applied
      avgIntensity: biasedAvgIntensity,
      biasPct: bias, // include bias metadata (non-breaking)
    };

    const header = `// ${workout.title} • FTP: ${workout.ftp} W • Bias: ${bias}%\n`;
    const dataStr = header + JSON.stringify(payload, null, 2);
    const dataBlob = new Blob([dataStr], { type: "application/json" });
    const url = URL.createObjectURL(dataBlob);

    const safeTitle = workout.title.replace(/[^a-zA-Z0-9]/g, "_");
    const link = document.createElement("a");
    link.href = url;
    link.download = `${safeTitle}_bias_${bias}.json`;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    URL.revokeObjectURL(url);

    toast({
      title: "Export successful",
      description: "Workout exported as JSON (with bias).",
    });
  };

  const handleExportText = () => {
    if (!workout) return;
    const data = buildWorkoutText();
    const blob = new Blob([data], { type: "text/plain" });
    const url = URL.createObjectURL(blob);
    const safeTitle = workout.title.replace(/[^a-zA-Z0-9]/g, "_");
    const link = document.createElement("a");
    link.href = url;
    link.download = `${safeTitle}.txt`;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    URL.revokeObjectURL(url);
  };

  const handleExportZWO = () => {
    if (!workout) return;
    const xml = toZwoXml({ ...workout, biasPct: bias });
    const blob = new Blob([xml], { type: "text/xml" });
    const url = URL.createObjectURL(blob);
    const safeTitle = workout.title.replace(/[^a-zA-Z0-9]/g, "_");
    const link = document.createElement("a");
    link.href = url;
    link.download = `${safeTitle}.zwo`;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    URL.revokeObjectURL(url);

    toast({
      title: "Export successful",
      description: "Workout exported as ZWO.",
    });
  };

  // Copy Text button removed in new UX

  // step list visuals removed in favor of the chart

  return (
    <div className="rounded-2xl bg-[--card] border border-[--border] p-6 shadow-[--shadow-card]">
      <div className="flex items-center mb-5">
        <ListOrdered className="text-[--accent-solid] mr-3 h-5 w-5" />
        <h2 className="text-xl font-semibold text-[--text-primary]">
          Generated Workout
        </h2>
      </div>

      {/* Bias controls */}
      {workout && (
        <div className="mb-5">
          {/* Top row: label + current value */}
          <div className="flex items-center justify-between mb-2">
            <div className="flex items-center gap-1">
              <span className="text-xs text-[--text-tertiary] uppercase tracking-wider">
                Bias
              </span>
              <TooltipProvider>
                <Tooltip>
                  <TooltipTrigger asChild>
                    <Info className="h-3.5 w-3.5 text-[--text-tertiary] cursor-pointer" />
                  </TooltipTrigger>
                  <TooltipContent
                    side="top"
                    className="max-w-[240px] text-sm rounded-md px-3 py-2 bg-[--card] text-[--text-primary] border border-[--border] shadow-lg"
                  >
                    Adjusts workout intensity on the fly. <br />
                    100% = planned watts, <br />
                    lower = easier, <br />
                    higher = harder.
                  </TooltipContent>
                </Tooltip>
              </TooltipProvider>
            </div>
            <span className="text-sm text-[--text-secondary] tabular-nums">
              {bias}%
            </span>
          </div>

          {/* Bottom row: - slider + */}
          <div className="flex items-center gap-2">
            <Button
              variant="outline"
              size="sm"
              className="h-11 w-11 rounded-full inline-flex items-center justify-center rounded-2xl px-2 py-1 font-medium transition-colors duration-150 focus:outline-none focus:ring-2 focus:ring-offset-0 focus:ring-emerald-500/60 bg-[--muted] text-[--text-secondary] hover:bg-[--border]"
              onClick={() => nudge(-1)}
              aria-label="Decrease bias"
              data-testid="bias-dec"
            >
              <Minus className="h-4 w-4" />
            </Button>

            <input
              type="range"
              min={BIAS_MIN}
              max={BIAS_MAX}
              step={1}
              value={bias}
              onChange={(e) => setBias(parseInt(e.target.value, 10))}
              className="flex-1 h-2 accent-emerald-500"
              aria-label="Bias percentage"
              data-testid="bias-range"
            />

            <Button
              variant="outline"
              size="sm"
              className="h-11 w-11 rounded-full inline-flex items-center justify-center rounded-2xl px-2 py-1 font-medium transition-colors duration-150 focus:outline-none focus:ring-2 focus:ring-offset-0 focus:ring-emerald-500/60 bg-[--muted] text-[--text-secondary] hover:bg-[--border]"
              onClick={() => nudge(1)}
              aria-label="Increase bias"
              data-testid="bias-inc"
            >
              <Plus className="h-4 w-4" />
            </Button>
          </div>
        </div>
      )}

      {workout ? (
        <div className="workout-content" data-testid="workout-display">
          {/* Workout Title */}
          <div className="mb-6">
            <h3
              className="text-xl font-semibold text-[--text-primary]"
              data-testid="text-workout-title"
            >
              {workout.title}
            </h3>
          </div>

          {/* Workout Chart (biased view) */}
          <WorkoutChart steps={biasedSteps} ftp={workout.ftp} />

          {/* Workout Summary (avg in biased W) */}
          <div className="mt-6 pt-6 border-t border-[--border]">
            <div className="grid grid-cols-2 sm:grid-cols-4 gap-4 text-center">
              <div className="bg-[--muted]/30 rounded-lg p-3">
                <div
                  className="text-2xl font-bold text-[--text-primary] tabular-nums"
                  data-testid="text-total-minutes"
                >
                  {workout.totalMinutes}'
                </div>
                <div className="text-xs text-[--text-tertiary] uppercase tracking-wider">
                  Total Time
                </div>
              </div>
              <div className="bg-[--muted]/30 rounded-lg p-3">
                <div
                  className="text-2xl font-bold text-[--text-primary] tabular-nums"
                  data-testid="text-work-minutes"
                >
                  {workout.workMinutes || 0}'
                </div>
                <div className="text-xs text-[--text-tertiary] uppercase tracking-wider">
                  Work Time
                </div>
              </div>
              <div className="bg-[--muted]/30 rounded-lg p-3">
                <div
                  className="text-2xl font-bold text-[--text-primary] tabular-nums"
                  data-testid="text-recovery-minutes"
                >
                  {workout.recoveryMinutes || 0}'
                </div>
                <div className="text-xs text-[--text-tertiary] uppercase tracking-wider">
                  Recovery
                </div>
              </div>
              <div className="bg-[--muted]/30 rounded-lg p-3">
                <div
                  className="text-2xl font-bold text-[--text-primary] tabular-nums"
                  data-testid="text-avg-intensity"
                >
                  {biasedAvgIntensity}W
                </div>
                <div className="text-xs text-[--text-tertiary] uppercase tracking-wider">
                  Avg Intensity
                </div>
              </div>
            </div>
          </div>

          {/* Export Actions */}
          <div className="mt-6 pt-6 border-t border-[--border]">
            <div className="flex flex-col sm:flex-row sm:flex-wrap gap-3">
              <DropdownMenu open={menuOpen} onOpenChange={setMenuOpen}>
                <DropdownMenuTrigger asChild>
                  <Button
                    disabled={!workout}
                    className="w-full sm:flex-1 min-w-0 inline-flex items-center justify-center gap-2 rounded-2xl font-medium transition-colors duration-150 focus:outline-none focus:ring-2 focus:ring-offset-0 focus:ring-emerald-500/60 bg-[--accent-solid] text-[--text-primary] hover:bg-[--accent-solidHover] border-[--text-secondary]"
                    aria-haspopup="menu"
                    aria-expanded={menuOpen}
                    aria-label="Open export menu"
                    data-testid="button-export-dropdown"
                  >
                    <Download className="h-4 w-4" />
                    Export
                    <ChevronDown className="h-4 w-4 ml-2" />
                  </Button>
                </DropdownMenuTrigger>
                <DropdownMenuContent
                  role="menu"
                  className="bg-[--muted] text-[--text-primary] border border-[--border] rounded-lg shadow-lg min-w-[180px] w-auto animate-in fade-in slide-in-from-top-1 duration-150 ease-out data-[state=closed]:animate-out data-[state=closed]:fade-out data-[state=closed]:slide-out-to-top-1 "
                  align="start"
                  sideOffset={8}
                >
                  <DropdownMenuItem
                    onClick={handleExportJSON}
                    role="menuitem"
                    aria-label="Export workout as JSON"
                    data-testid="menu-export-json"
                    className="px-3 py-2.5 flex items-center gap-2 hover:bg-[--accent-solid]/10 rounded-md text-[--text-primary] cursor-pointer"
                  >
                    <Code className="h-4 w-4" />
                    <span>Export JSON</span>
                  </DropdownMenuItem>

                  <DropdownMenuItem
                    onClick={handleExportText}
                    role="menuitem"
                    aria-label="Export workout as Text"
                    data-testid="menu-export-text"
                    className="px-3 py-2.5 flex items-center gap-2 hover:bg-[--accent-solid]/10 rounded-md text-[--text-primary] cursor-pointer"
                  >
                    <FileText className="h-4 w-4" />
                    <span>Export Text</span>
                  </DropdownMenuItem>
                  <DropdownMenuItem
                    onClick={handleExportZWO}
                    role="menuitem"
                    aria-label="Export workout as ZWO"
                    data-testid="menu-export-zwo"
                    className="px-3 py-2.5 flex items-center gap-2 hover:bg-[--accent-solid]/10 rounded-md text-[--text-primary] cursor-pointer"
                  >
                    <FileCog className="h-4 w-4" />
                    <span>Export ZWO</span>
                  </DropdownMenuItem>
                </DropdownMenuContent>
              </DropdownMenu>
            </div>
          </div>
        </div>
      ) : attempted ? (
        <div className="empty-state text-center py-12" data-testid="empty-state">
          <Bike className="mx-auto text-4xl text-[--text-tertiary] mb-4 h-16 w-16" />
          <h3 className="text-lg font-medium text-[--text-secondary] mb-2">
            No workout found
          </h3>
          <p className="text-[--text-tertiary]">
            No pattern fits the selected duration range. Try a longer range or another type.
          </p>
        </div>
      ) : (
        <div className="empty-state text-center py-12" data-testid="empty-state">
          <Bike className="mx-auto text-4xl text-[--text-tertiary] mb-4 h-16 w-16" />
          <h3 className="text-lg font-medium text-[--text-secondary] mb-2">
            No workout generated yet
          </h3>
          <p className="text-[--text-tertiary]">
            Configure your settings and click "Generate Workout" to create a personalized training session.
          </p>
        </div>
      )}
    </div>
  );
}
