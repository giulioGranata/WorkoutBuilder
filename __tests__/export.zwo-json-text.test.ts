import { describe, it, expect } from "vitest";
import { toZwoXml } from "@/lib/zwo";
import { applyBias } from "@/components/WorkoutOutput";
import type { Workout, Step } from "@/lib/types";

const workout: Workout = {
  title: "Export Sample",
  ftp: 200,
  steps: [
    { kind: "steady", minutes: 5, intensity: 150, description: "steady", phase: "work" },
    { kind: "ramp", minutes: 5, from: 100, to: 140, description: "ramp", phase: "work" },
  ],
  totalMinutes: 10,
  workMinutes: 10,
  recoveryMinutes: 0,
  avgIntensity: 150,
  signature: "sig",
};

function buildJson(w: Workout, bias: number) {
  const steps: Step[] = w.steps.map((s) => {
    const kind = (s as any).kind ?? "steady";
    if (kind === "ramp") {
      const r = s as any;
      return { ...r, from: applyBias(r.from, bias), to: applyBias(r.to, bias), kind: "ramp" };
    }
    const st = s as any;
    return { ...st, intensity: applyBias(st.intensity, bias), kind: "steady" };
  });
  const avg = Math.round(
    steps.reduce((sum, s: any) => {
      if (s.kind === "ramp") {
        return sum + ((s.from + s.to) / 2) * s.minutes;
      }
      return sum + s.intensity * s.minutes;
    }, 0) / w.totalMinutes
  );
  const payload = { ...w, steps, avgIntensity: avg, biasPct: bias };
  const header = `// ${w.title} • FTP: ${w.ftp} W • Bias: ${bias}%\n`;
  return header + JSON.stringify(payload, null, 2);
}

function buildText(w: Workout, bias: number) {
  const steps = w.steps.map((s) => {
    const kind = (s as any).kind ?? "steady";
    if (kind === "ramp") {
      const r = s as any;
      return { ...r, from: applyBias(r.from, bias), to: applyBias(r.to, bias), kind: "ramp" };
    }
    const st = s as any;
    return { ...st, intensity: applyBias(st.intensity, bias), kind: "steady" };
  });
  const avg = Math.round(
    steps.reduce((sum, s: any) => {
      if (s.kind === "ramp") {
        return sum + ((s.from + s.to) / 2) * s.minutes;
      }
      return sum + s.intensity * s.minutes;
    }, 0) / w.totalMinutes
  );
  const header = `${w.title} • FTP: ${w.ftp} W • Bias: ${bias}%\n\n`;
  const body = steps
    .map((step, i) => {
      const kind = step.kind;
      const wattsText =
        kind === "ramp"
          ? `ramp ${(step as any).from}→${(step as any).to} W`
          : `${(step as any).intensity} W`;
      return `${i + 1}. ${step.minutes}' — ${wattsText} — ${step.description}`;
    })
    .join("\n");
  const footer = `\n\nTotal: ${w.totalMinutes}'\nAvg: ${avg} W`;
  return header + body + footer;
}

describe("export formats", () => {
  it("produces consistent ZWO, JSON and Text outputs", () => {
    const bias = 110;
    const xml = toZwoXml({ ...workout, biasPct: bias });
    const json = buildJson(workout, bias);
    const text = buildText(workout, bias);

    expect(xml).toMatchSnapshot("zwo");
    expect(json).toMatchSnapshot("json");
    expect(text).toMatchSnapshot("text");

    const jsonStr = json.replace(/^\/\/.*\n/, "");
    expect(() => JSON.parse(jsonStr)).not.toThrow();
  });
});
