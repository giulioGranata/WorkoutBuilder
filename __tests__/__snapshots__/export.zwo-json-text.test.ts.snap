// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`export formats > produces consistent ZWO, JSON and Text outputs > json 1`] = `
"// Export Sample • FTP: 200 W • Bias: 110%
{
  "title": "Export Sample",
  "ftp": 200,
  "steps": [
    {
      "kind": "steady",
      "minutes": 5,
      "intensity": 165,
      "description": "steady",
      "phase": "work"
    },
    {
      "kind": "ramp",
      "minutes": 5,
      "from": 110,
      "to": 154,
      "description": "ramp",
      "phase": "work"
    }
  ],
  "totalMinutes": 10,
  "workMinutes": 10,
  "recoveryMinutes": 0,
  "avgIntensity": 149,
  "signature": "sig",
  "biasPct": 110
}"
`;

exports[`export formats > produces consistent ZWO, JSON and Text outputs > text 1`] = `
"Export Sample • FTP: 200 W • Bias: 110%

1. 5' — 165 W — steady
2. 5' — ramp 110→154 W — ramp

Total: 10'
Avg: 149 W"
`;

exports[`export formats > produces consistent ZWO, JSON and Text outputs > zwo 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<workout_file>
  <author>Workout Generator 2.0</author>
  <name>Export Sample</name>
  <description>FTP 200W • Bias 110%</description>
  <workout>
    <SteadyState Duration="300" Power="0.75" />
    <Ramp Duration="300" PowerLow="0.50" PowerHigh="0.70" />
  </workout>
</workout_file>"
`;
